skcosmo/pcovr/pcovr.py

```diff
@@ -39,16 +46,37 @@ def pcovr_covariance(mixing, X, Y, rcond=1e-12, return_isqrt=False):
                       defaults to 1E-12
     :type rcond: float
 
+    :param return_isqrt: Whether to return the calculated inverse square root of
+                         the covariance. Used when inverse square root is needed
+                         and the pcovr_covariance has already been calculated
+    :type return_isqrt: boolean
+
+    :param n_components: number of eigenpairs to estimate the inverse square root
+                         with. Defaults to min(X.shape)
+    :type n_components: int
+
+    :param random_state: random seed to use for randomized svd
+    :type random_state: int
+
     """
 
     C = np.zeros((X.shape[1], X.shape[1]), dtype=np.float64)
 
+    if n_components is None:
+        n_components = min(X.shape)
+
     cov = X.T @ X
 
     if mixing < 1 or return_isqrt:
-        # Do not try to approximate C_inv, it will affect results
-        C_inv = np.linalg.pinv(cov, rcond=rcond)
-        C_isqrt = np.real(scipy.linalg.sqrtm(C_inv))
+
+        _, vC, UC = randomized_svd(
```

Comment:
So here, because we're inverting the eigenvalues, we are almost never going to be wanting the full eigendecomposition, otherwise the square root inverse will be unphysically large. From the sklearn options, this eliminates "full" as a reasonable choice. I can modify it to consider arpack as well, although I chose randomized here because it is the sklearn default for truncated SVDs. Do you think it's crucial to have both?

Reply:
But couldn't you do the same argument for the svd PCovR?
Also in the PCovR class in `_fit_feature_space`, the `n_components` is not an input argument for the function `pcovr_covariance`, so the PCA svd solver would theoretically default to the full method in this case.

Reply:
You want that -- the n_components used to construct Ctilde should not be the same as the n_components used for the projection. Maybe I should rename the n_components in constructing Ctilde? 

Essentially, it is incredibly costly to compute the inverse square root through usual methods (scipy.linalg.sqrtm), so an approximation of Ctilde will suffice using the eigendecomposition. _However_ this approximation shouldn't be low-rank like the PCA, but rather high enough rank to correctly construct Ctilde. I called it n_components to be consistent with the randomized_svd naming, but if it leads to confusion with the n_PCA components, it might be worth changing.

Reply:
I've renamed it `rank` to avoid this confusion. It should infer the rank that you assume the covariance is.